src/shared/openssl-util.c

```diff
@@ -269,6 +279,177 @@ int string_hashsum(
 
 }
 #  endif
+
+/* We use derived shared secret as volume key for EC keys. The following is the procedure. First we generate
+ * a pair of EC key of the same curve group. Then we use newly generated private key to derive shared secret
+ * with the public key in the token. Finally, we simplely forget the private key, store the public key to
+ * pkcs11-key field of the LUKS2 header and use the derived shared secret as the volume key (we don't apply
+ * any KDF here since LUKS will apply PBKDF on the key).
+ */
+static int pkey_generate_volume_key_ec(
+                EVP_PKEY *pkey,
+                void **ret_decrypted_key,
+                size_t *ret_decrypted_key_size,
+                void **ret_saved_key,
+                size_t *ret_saved_key_size) {
+
+        _cleanup_(EVP_PKEY_freep) EVP_PKEY *pkey_new = NULL;
+        _cleanup_free_ uint8_t *decrypted_key = NULL;
```

Comment:
this should use `_cleanup_(erase_and_freep)`, right? i.e. it's sensitive data, no?